pulp_python/app/migrations/0019_create_missing_metadata_artifacts.py

```diff
@@ -0,0 +1,188 @@
+                    return f.read(file_path)
+    except (zipfile.BadZipFile, KeyError, OSError):
+        pass
+    return None
+
+
+def artifact_to_metadata_artifact(filename, artifact, md_digests, tmp_dir, artifact_model):
+    """
+    Creates artifact for metadata from the provided wheel artifact.
+    """
+    import shutil
+    import tempfile
+
+    with tempfile.NamedTemporaryFile("wb", dir=tmp_dir, suffix=filename, delete=False) as temp_file:
+        temp_wheel_path = temp_file.name
+        artifact.file.seek(0)
+        shutil.copyfileobj(artifact.file, temp_file)
+        temp_file.flush()
+
+    metadata_content = extract_wheel_metadata(temp_wheel_path)
+    if not metadata_content:
+        return None
+
+    with tempfile.NamedTemporaryFile(
+        "wb", dir=tmp_dir, suffix=".metadata", delete=False
+    ) as temp_md:
+        temp_metadata_path = temp_md.name
+        temp_md.write(metadata_content)
+        temp_md.flush()
+
+    metadata_artifact = init_and_validate(temp_metadata_path, artifact_model, md_digests)
+    return metadata_artifact
+
+
+def create_missing_metadata_artifacts(apps, schema_editor):
+    """
+    Create metadata artifacts for PythonPackageContent instances that have metadata_sha256
+    but are missing the corresponding metadata artifact.
+    """
+    import tempfile
+    from django.conf import settings
+
+    PythonPackageContent = apps.get_model("python", "PythonPackageContent")
+    ContentArtifact = apps.get_model("core", "ContentArtifact")
+    Artifact = apps.get_model("core", "Artifact")
+
+    packages = (
+        PythonPackageContent.objects.filter(
+            metadata_sha256__isnull=False, filename__endswith=".whl"
+        )
+        .exclude(metadata_sha256="")
+        .prefetch_related("contentartifact_set")
+        .only("filename", "metadata_sha256")
+    )
+    artifact_batch = []
+    contentartifact_batch = []
+
+    with tempfile.TemporaryDirectory(dir=settings.WORKING_DIRECTORY) as temp_dir:
+        for package in packages:
+            filename = package.filename
+            content_artifacts = list(package.contentartifact_set.all())
+
+            # Get the main artifact for package
+            main_artifact = None
+            for ca in content_artifacts:
+                if ca.relative_path == filename and ca.artifact:
+                    main_artifact = ca.artifact
+                    break
+
+            if not main_artifact:
+                # Main artifact does not exist
+                continue
+
+            metadata_digests = {"sha256": package.metadata_sha256}
+            metadata_artifact = artifact_to_metadata_artifact(
+                filename, main_artifact, metadata_digests, temp_dir, Artifact
+            )
+            if not metadata_artifact:
+                # Failed to build metadata artifact
+                continue
+
+            contentartifact = ContentArtifact(
+                artifact=metadata_artifact,
```

Comment:
I don't think this will work. The metadata_artifact is just an instance in memory, it doesn't have a pk needed for this foreign key, and I don't believe that bulk_create goes and updates the passed in instances.